modules/partition-replicator/src/test/java/org/apache/ignite/internal/partition/replicator/NaiveAsyncReadWriteLockTest.java

```diff
@@ -0,0 +1,120 @@
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ForkJoinPool;
+import org.apache.ignite.internal.testframework.ExecutorServiceExtension;
+import org.apache.ignite.internal.testframework.InjectExecutorService;
+import org.apache.ignite.internal.util.CompletableFutures;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@ExtendWith(ExecutorServiceExtension.class)
+class NaiveAsyncReadWriteLockTest {
+    private final NaiveAsyncReadWriteLock lock = new NaiveAsyncReadWriteLock(ForkJoinPool.commonPool());
+
+    @Test
+    void writeLockAttemptsArrivingWhenWriteLockedGetSatisfiedWhenUnlocked() {
+        List<CompletableFuture<?>> futures = new ArrayList<>();
+
+        CompletableFuture<Long> firstFuture = lock.writeLock();
+
+        for (int i = 0; i < 10; i++) {
+            futures.add(lock.writeLock().thenAccept(lock::unlockWrite));
+        }
+
+        assertThat(firstFuture.thenAccept(lock::unlockWrite), willCompleteSuccessfully());
+
+        assertThat(CompletableFutures.allOf(futures), willCompleteSuccessfully());
+    }
+
+    @Test
+    void readLockAttemptsArrivingWhenWriteLockedGetSatisfiedWhenUnlocked() {
+        List<CompletableFuture<?>> futures = new ArrayList<>();
+
+        CompletableFuture<Long> firstFuture = lock.writeLock();
+
+        for (int i = 0; i < 10; i++) {
+            futures.add(lock.readLock().thenAccept(lock::unlockRead));
+        }
+
+        assertThat(firstFuture.thenAccept(lock::unlockWrite), willCompleteSuccessfully());
+
+        assertThat(CompletableFutures.allOf(futures), willCompleteSuccessfully());
+    }
+
+    @Test
+    void writeLockAttemptsArrivingWhenReadLockedGetSatisfiedWhenUnlocked() {
+        CompletableFuture<Long> readLockFuture1 = lock.readLock();
+        CompletableFuture<Long> readLockFuture2 = lock.readLock();
+
+        CompletableFuture<Long> writeLockFuture = lock.writeLock();
+
+        assertThat(readLockFuture1.thenAccept(lock::unlockRead), willCompleteSuccessfully());
+        assertThat(readLockFuture2.thenAccept(lock::unlockRead), willCompleteSuccessfully());
+
+        assertThat(writeLockFuture, willCompleteSuccessfully());
+        assertDoesNotThrow(() -> lock.unlockWrite(writeLockFuture.join()));
+    }
+
+    @Test
+    void testConcurrency(@InjectExecutorService(threadCount = 4) ExecutorService executor) {
+        List<CompletableFuture<?>> futures = new CopyOnWriteArrayList<>();
+
+        for (int thread = 0; thread < 4; thread++) {
+            executor.execute(() -> {
+                for (int i = 0; i < 10000; i++) {
+                    CompletableFuture<?> future;
+
+                    if (i % 2 == 0) {
+                        future = lock.readLock().thenCompose(stamp -> {
+                            return CompletableFuture.runAsync(() -> sleep(1))
```

Comment:
Should this stuff be submitted on an explicit executor? This can drain the default pool otherwise